templates/examples/extensions/py_lambda_hooks/CanvasHook/CanvasHook.py

```diff
@@ -0,0 +1,405 @@
+    course_syllabus = ''
+
+    # Get the user using user_id to match with LMS SIS_ID
+    user = getCanvasUser (canvas, student_user_name)
+    if user:
+        course_list = []
+        courses = user.get_courses(enrollment_status='active',include=['syllabus_body'])
+        
+        for course in courses:
+            course_list.append(course.name)
+        choice = process.extractOne(userinput, course_list, scorer=fuzz.token_set_ratio)
+
+        if courses: 
+            # Loop through the courses.
+            for course in courses:
+                #check for matching userinput with course names
+                if course.name == choice[0] and choice[1] > MATCHING_TOLERANCE_SCORE:
+                    if course.syllabus_body.strip() != '':
+                        course_syllabus = '<b>{0}</b>: {1}<br>'.format(course.name, course.syllabus_body)
+                        break
+                    else:
+                        course_syllabus = no_syllabus
+                        break
+                else:
+                    if course.syllabus_body.strip() != '':
+                        course_syllabus += '<b>{0}</b>: {1}<br>'.format(course.name, course.syllabus_body)
+                    else:
+                        course_syllabus += '<b>{0}</b>: {1}<br>'.format(course.name, no_syllabus)
+
+    result = {"CourseSyllabus": course_syllabus}
+    return result
+
+
+#----------------------------------------------------------------------
+# function: validate_input
+#----------------------------------------------------------------------
+def validate_input(event):
+    error_message = ''
+
+    try:
+        if not error_message:
+            error_message = 'You must provide a question.'
+            if json.loads(event['res']['result']['args'][0])['Query']:
+                error_message = ''
+
+        if not error_message:
+            error_message = 'You are not logged in.'
+            if event['req']['_userInfo']['Email']:
+                error_message = ''
+    except:
+        pass
+
+    return error_message
+
+
+#----------------------------------------------------------------------
+# function handler
+#----------------------------------------------------------------------
+def handler(event, context):
+    userinput = event["req"]["_event"]["inputTranscript"]
+    return_message = ''
+
+    # Validate the required input.
+    error_message = validate_input(event)
+
+    if error_message:
+        return_message = error_message
+        #event['res']['message'] = return_message
+        event['res']['session']['appContext']['altMessages']['markdown'] = return_message
+    else:
+        # Get the API domain. This will be need for API calls and for looking up the bearer token.
+        domain = event['req']['_settings']['CanvasDomainName']
+        secrets_id_name = event['req']['_settings']['CanvasAPIKey']
+
+        # Get the bearer token from Secrets Manager.
+        api_token = get_secret(secrets_id_name)
```

Comment:
Should api_token be cached in lambda and called outside the lambda handler?

Possibly canvas could be cached if its constructor is establishing a connection/session to canvas and we want to optimize by re-using the canvas connection across many lambda invocations. Would that be worthwhile?